libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -297,6 +303,9 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 	if (_variable.annotation().type)
 		return;
 
+	if (_variable.isConstant() && !_variable.isStateVariable())
+		m_errorReporter.declarationError(_variable.location(), "The \"constant\" keyword can only be used for state variables.");
```

Comment:
Is this actually part of this PR?

Reply:
Previously, the checks were much less restrictive, so the checks would pass here and then the type checker did the rest, but I actually think it is better to have strict tests here and then also move this check here, which I would say is the appropriate place.

Reply:
Ok perhaps to summarize differently: Other ways to solve this would result in weird error messages.

Reply:
If you agree, I would still like to keep it here. While it can be checked earlier, I think it still "belongs" here, especially when we also add things like `mutable` and `immutable`.

Reply:
I'm fine with keeping it here.